compiler/daml-lf-proto/src/DA/Daml/LF/Mangling.hs

```diff
@@ -108,39 +109,42 @@ mangleIdentifier txt = case T.foldl' f (MangledSize 0 0) txt of
-  chs <- goStart (T.unpack txt)
-  if null chs
+  chs <- goStart txt
+  if T.null chs
     then Left "Empty identifier"
-    else Right (T.pack chs)
+    else Right chs
   where
-    go :: (Char -> Bool) -> (String -> Either String String) -> String -> Either String String
-    go isAllowed followUp = \case
-      [] -> Right []
-      ['$'] -> Left "Got control character $ with nothing after it. It should be followed by $, u, or U"
-      '$' : ctrlCh : chs0 -> case ctrlCh of
-        '$' -> ('$' :) <$> followUp chs0
-        'u' -> do
-          (ch, chs1) <- readEscaped "$u" 4 chs0
-          (ch :) <$> followUp chs1
-        'U' -> do
-          (ch, chs1) <- readEscaped "$U" 8 chs0
-          (ch :) <$> followUp chs1
-        ch -> Left ("Control character $ should be followed by $, u, or U, but got " ++ show ch)
-      ch : chs -> if isAllowed ch
-        then (ch :) <$> followUp chs
-        else Left ("Unexpected unescaped character " ++ show ch)
-    
+    go :: (Char -> Bool) -> (T.Text -> Either String T.Text) -> T.Text -> Either String T.Text
+    go isAllowed followUp s
+        | T.null s = Right T.empty
+        | Just ('$', s) <- T.uncons s =
+              case T.uncons s of
+                  Just ('$', s) -> T.cons '$' <$> followUp s
+                  Just ('u', s) -> do
+                      (ch, s') <- readEscaped "$u" 4 s
+                      T.cons ch <$> followUp s'
+                  Just ('U', s) -> do
+                      (ch, s') <- readEscaped "$U" 8 s
+                      T.cons ch <$> followUp s'
+                  _ -> Left "Control character $ should be followed by $, u or U"
+        | otherwise = case T.span isAllowed s of
+              (prefix, suffix)
+                  | T.null prefix -> Left ("Unexpected unescaped character " ++ show (T.head suffix))
+                  | otherwise -> fmap (prefix <>) (followUp suffix)
+
     goStart = go isAllowedStart goPart
     goPart = go isAllowedPart goPart
 
     readEscaped what n chs0 = let
-      (escaped, chs) = splitAt n chs0
+      (escaped, chs) = T.splitAt n chs0
       in if
-        | length escaped < n ->
-            Left ("Expected " ++ show n ++ " characters after " ++ what ++ ", but got " ++ show (length escaped))
-        | not (all (`Set.member` escapeSequencesChars) escaped) ->
+        | T.length escaped < n ->
+            Left ("Expected " ++ show n ++ " characters after " ++ what ++ ", but got " ++ show (T.length escaped))
+        | not (T.all isEscapeSequenceChars escaped) ->
             Left ("Expected only lowercase hex code in escape sequences, but got " ++ show escaped)
-        | otherwise -> Right (toEnum (read ("0x" ++ escaped)), chs)
+        | otherwise ->
+          -- We’ve already done the validation so fromRight is safe.
+          Right (toEnum (fst $ fromRight (error $ "Internal error in unmangleIdentifier: " <> show escaped) $ T.hexadecimal escaped), chs)
 
     -- only lowercase, as per printf
```

Comment:
Do we really need this restriction? What happens if we make the decoder more lenient? This would allow us to avoid the check above and fully rely on `T.hexadecimal`.

Reply:
Should be fine but I don’t want to add more things to this PR than necessary. I also don’t really expect it to make a difference.